pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -118,31 +120,147 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
-			services = append(services, service)
-			// If servicePorts has this service, that means it is old service.
-			if servicePorts[service.Hostname.String()] != nil {
-				oldPorts := servicePorts[service.Hostname.String()]
-				for port, cluster := range oldPorts {
-					// if this service port is removed, we can conclude that it is a removed cluster.
-					if _, exists := service.Ports.GetByPort(port); !exists {
-						deletedClusters = append(deletedClusters, cluster)
-					}
+		}
+	} else {
+		services = append(services, service)
+		// If servicePorts has this service, that means it is old service.
+		if servicePorts[service.Hostname.String()] != nil {
+			oldPorts := servicePorts[service.Hostname.String()]
+			for port, cluster := range oldPorts {
+				// if this service port is removed, we can conclude that it is a removed cluster.
+				if _, exists := service.Ports.GetByPort(port); !exists {
+					deletedClusters = append(deletedClusters, cluster)
 				}
 			}
 		}
 	}
-	clusters, log := configgen.buildClusters(proxy, updates, services)
-	return clusters, deletedClusters, log, true
+	return services, deletedClusters
+}
+
+// deltaFromDestinationRules computes the delta clusters from the updated destination rules.
+func (configgen *ConfigGeneratorImpl) deltaFromDestinationRules(updatedDr model.ConfigKey, proxy *model.Proxy, push *model.PushContext,
+	watched []string,
+) ([]*model.Service, []string) {
+	deletedClusters := make([]string, 0)
+	services := make([]*model.Service, 0)
+	cfg := push.DestinationRuleByName(proxy, updatedDr.Name, updatedDr.Namespace)
+	if cfg == nil {
+		// subset clusters need to be removed, actual cluster needs to be updated
+		// a destinationrule was deleted, get the populated destinationRule from PrevSidecarScope
+		prevCfg := push.PrevDestinationRuleByName(proxy, updatedDr.Name, updatedDr.Namespace)
+		if prevCfg == nil {
+			// something went wrong, skip -- we shouldn't be nil here
+			log.Warnf("Prev DestinationRule form PrevSidecarScope is nil for %v", updatedDr.String())
+			return nil, nil
+		}
+		// populate removed clusters ONLY for subsets.
+		dr := prevCfg.Spec.(*networking.DestinationRule)
+		deletedClusters = append(deletedClusters, removedClusterNames(subsetClusters(watched), nil, dr.Host)...)
+
+		// for non-subset destinationrule deletion -- generate the cluster as it changed
+		matched := push.ServicesForHostname(proxy, host.Name(dr.Host), prevCfg.Namespace)
+		services = append(services, matched...)
+	} else {
+		// destination exists, was updated
+		// generate cluster based on service associated with destinationrule
+		// hostName can be a wildcard, and can match to multiple services
+		dr := cfg.Spec.(*networking.DestinationRule)
+		matchedSvcs := push.ServicesForHostname(proxy, host.Name(dr.Host), cfg.Meta.Namespace)
+		services = append(services, matchedSvcs...)
+		// check for removed subsets
+		prevCfg := push.PrevDestinationRuleByName(proxy, updatedDr.Name, updatedDr.Namespace)
```

Comment:
is the usage of PrevDestinationRuleByName sound? I am not sure we guarantee we will push each push context?

For example, consider I create push contexts v1,v2,v3,v4. I a may end up only pushing v1 and v4 to a specific proxy, since they are deduped in PushQueue. So we may have the DR removed in v2, by the time we get to v4 the DR is not in PushContext nor prev pushcontext?

Reply:
I don't think we need it, either. We can just find all clusters that match the DR (from serviceClusters, although that only has exact hostnames but DR can do wildcard so we need some more logic) and remove all the subsets from there? We don't need to know the subsets from the previous DR, since we know all the known-clusters so we can just remove from that list

Reply:
Actually not sure my proposed logic or current logic works. Imagine if I rename a DR. We are both adding and deleting the same cluster in this case, which will not be handled correctly.

maybe that is solved by removing anything built clusters from `deletedClusters`

Reply:
I think in all cases, if DR is updated/removed just wipe off subset clusters (add it to deletedClusters) and reconcile with the new built clusters is lot simpler. I have changed it like that. PTAL

Reply:
I think any PrevDestinationRuleByName usage is probably wrong. Do we even need it now that we have the "remove all subset"?

Reply:
Seems like we still do actually since we don't even know the hostname unfortunately.

Reply:
correct